homeassistant/components/vacuum/__init__.py

```diff
@@ -0,0 +1,287 @@
+    SERVICE_TURN_ON, SERVICE_TURN_OFF, SERVICE_TOGGLE)
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.config_validation import PLATFORM_SCHEMA  # noqa
+from homeassistant.helpers.entity_component import EntityComponent
+from homeassistant.helpers.entity import ToggleEntity
+from homeassistant.util.icon import icon_for_battery_level
+
+_LOGGER = logging.getLogger(__name__)
+
+DOMAIN = 'vacuum'
+
+SCAN_INTERVAL = timedelta(seconds=20)
+
+GROUP_NAME_ALL_VACUUMS = 'all vacuum cleaners'
+
+ATTR_COMMAND = 'command'
+ATTR_FANSPEED = 'fanspeed'
+ATTR_PARAMS = 'params'
+
+SERVICE_START_PAUSE = 'start_pause'
+SERVICE_LOCATE = 'locate'
+SERVICE_RETURN_TO_BASE = 'return_to_base'
+SERVICE_SEND_COMMAND = 'send_command'
+SERVICE_SET_FANSPEED = 'set_fanspeed'
+SERVICE_STOP = 'stop'
+
+VACUUM_SERVICE_SCHEMA = vol.Schema({
+    vol.Optional(ATTR_ENTITY_ID): cv.entity_ids,
+})
+
+VACUUM_SET_FANSPEED_SERVICE_SCHEMA = VACUUM_SERVICE_SCHEMA.extend({
+    vol.Required(ATTR_FANSPEED): cv.string,
+})
+
+VACUUM_SEND_COMMAND_SERVICE_SCHEMA = VACUUM_SERVICE_SCHEMA.extend({
+    vol.Required(ATTR_COMMAND): cv.string,
+    vol.Optional(ATTR_PARAMS): vol.Any,
+})
+
+SERVICE_TO_METHOD = {
+    SERVICE_TURN_ON: {'method': 'async_turn_on'},
+    SERVICE_TURN_OFF: {'method': 'async_turn_off'},
+    SERVICE_TOGGLE: {'method': 'async_toggle'},
+    SERVICE_START_PAUSE: {'method': 'async_start_pause'},
+    SERVICE_RETURN_TO_BASE: {'method': 'async_return_to_base'},
+    SERVICE_LOCATE: {'method': 'async_locate'},
+    SERVICE_STOP: {'method': 'async_stop'},
+    SERVICE_SET_FANSPEED: {'method': 'async_set_fanspeed',
+                           'schema': VACUUM_SET_FANSPEED_SERVICE_SCHEMA},
+    SERVICE_SEND_COMMAND: {'method': 'async_send_command',
+                           'schema': VACUUM_SEND_COMMAND_SERVICE_SCHEMA},
+}
+
+DEFAULT_NAME = 'Vacuum cleaner robot'
+DEFAULT_ICON = 'mdi:google-circles-group'
+
+SUPPORT_TURN_ON = 1
+SUPPORT_TURN_OFF = 2
+SUPPORT_PAUSE = 4
+SUPPORT_STOP = 8
+SUPPORT_RETURN_HOME = 16
+SUPPORT_FANSPEED = 32
+SUPPORT_BATTERY = 64
+SUPPORT_STATUS = 128
+SUPPORT_SENDCOMMAND = 256
+SUPPORT_LOCATE = 512
+SUPPORT_MAP = 1024
+
+
+@asyncio.coroutine
+def async_setup(hass, config):
+    """Set up the vacuum component."""
+    if not config[DOMAIN]:
+        return False
+
+    component = EntityComponent(
+        _LOGGER, DOMAIN, hass, SCAN_INTERVAL, GROUP_NAME_ALL_VACUUMS)
+
+    yield from component.async_setup(config)
+
+    descriptions = yield from hass.async_add_job(
+        load_yaml_config_file, os.path.join(
+            os.path.dirname(__file__), 'services.yaml'))
+
+    @asyncio.coroutine
+    def async_handle_vacuum_service(service):
+        """Map services to methods on VacuumDevice."""
+        method = SERVICE_TO_METHOD.get(service.service)
+        if not method:
+            return
+
+        target_vacuums = component.async_extract_from_service(service)
+        params = dict(service.data)
```

Comment:
sorry, I thought that `async_extract_from_service` was doing that. Fixed, and also getting the params as the `cover` does:
```python
params = service.data.copy()
params.pop(ATTR_ENTITY_ID, None)
```